wsd/DocumentBroker.hpp

```diff
@@ -540,6 +540,46 @@ class DocumentBroker : public std::enable_shared_from_this<DocumentBroker>
     /// a convert-to request or doctored to look like one.
     virtual bool isConvertTo() const { return false; }
 
+private:
+    /// Responsible for managing document saving.
+    /// Tracks auto-saveing and its frequency.
+    class SaveManager final
+    {
+    public:
+        SaveManager()
+            : _lastAutosaveCheckTime(std::chrono::steady_clock::now())
+            , _isAutosaveEnabled(std::getenv("LOOL_NO_AUTOSAVE") == nullptr)
+        {
+        }
+
+        /// Return true iff auto save is enabled.
+        bool isAutosaveEnabled() const { return _isAutosaveEnabled; }
+
+        /// Returns true if we should issue an auto-save.
+        bool needAutosaveCheck() const
+        {
+            return isAutosaveEnabled()
+                   && std::chrono::duration_cast<std::chrono::seconds>(now()
+                                                                       - _lastAutosaveCheckTime)
+                          >= std::chrono::seconds(30);
+        }
+
+        /// Marks autosave check done.
+        void autosaveChecked() { _lastAutosaveCheckTime = now(); }
```

Comment:
And setAutosaveChecked() maybe here?

Reply:
Actually in a later patch, you use "mar", as in markLastSaveResponseTime(), that would make sense here too maybe, like markAutosaveChecked()?